### ISSUE TITLE
feature request: swing - shuffle - groove

### ISSUE THREAD

**Issue by s-a** [mohayonao/ciseaux#15]
Here are some groove settings available in propellerhead reason. I think it is not too difficult to implement, but I do not know the logic behind their different groove patches.

![regroove-tool-window](https://cloud.githubusercontent.com/assets/1270662/6452898/18624e10-c139-11e4-9b91-309426143bee.jpg)

**Comment by mohayonao** [comment]
like this?

```js
// convert 1:1 to 2:1
tape = Ciseaux.concat(tape.split(32).map(function(tape, index) {
  return index % 2 ? tape.stretch(1 / 1.5) : tape.stretch(2 / 1.5);
}));
```

but, `Tape#stretch()` is not implemented yet. (very difficult)

**Comment by s-a** [comment]
I suggest to really call it shuffle or better groove and ignore the note length option for the first step which is firstly not important when shuffling beats . So stretch would be nice to have. A simple swing is easy to code and is based on a very tiny ``positive or negative tick offset```. I found [this topic] (http://stackoverflow.com/questions/11759459/audio-sequencer-with-swing-shuffle-obj-c) useful .
You will wonder how human it sounds when beats are shuffled only very slightly including some velocity variation.

    if( tick <= 16 )
        tick *= s; //stretch
    else
        tick = (2-s)*tick + 32*(s-1) //compress

This what we have to learn and define in the groove shuffle patches.

**Comment by mohayonao** [comment]
umm.. Like this? This is easy.

```js
// durationPerStep : 1 1 1 1 1 1
// sequence pattern: a b a b a b
tape = new Ciseaux.Sequence("ababab", 1)
  .apply( instruments );

// durationPerStep:  1 2   3     1  2  3
// sequence pattern: a b - a - - b a - b - - ...
tape = new Ciseaux.Sequence("ababab", [ 1, 2, 3 ])
  .apply( instruments );
```

**Comment by s-a** [comment]
Yes. I do not understand the code completely but the sequence pattern description looks good. 
I think about how to declare shuffle settings.  How to declare timing offset per step?

**Comment by mohayonao** [comment]
arguments description:

```js
// durationPerStep : 1 2   3     1 2   3
// sequence pattern: a b - a - - b a - b - -
tape = new Ciseaux.Sequence("ababab", [ 1, 2, 3 ])
  .apply({a:tape1, b:tape2}); // ^    ^
//       ^                       |    |
//       |                       |  durationPerStep
//       instruments             pattern

// that means
tape = Ciseaux.concat([
  tape1.slice(0, 1),  // a
  tape2.slice(0, 2),  // b
  tape1.slice(0, 3),  // a
  tape2.slice(0, 1),  // b
  tape1.slice(0, 2),  // a
  tape2.slice(0, 3),  // b
]);
```

random timing:

```js
var randomDurationPerStep = [ 1, 2, 3 ].map(function(durationPerStep) {
  return durationPerStep + Math.random() * 0.2 - 0.1;
});

tape = new Ciseaux.Sequence("ababab", ranromDurationPerStep)
  .apply({a:tape1, b:tape2});
```

random velocity:

```js
var randomVelocity = function(tape) {
  return tape.gain(Math.random() * 0.5 + 0.5);
};

tape = new Ciseaux.Sequence("ababab", [ 1, 2, 3 ])
  .apply({
    // instruments can accept a function that returns a tape
    a: function() { return randomVelocity(tape1); },
    b: function() { return randomVelocity(tape2); },
  });
```

and, arguments are flexible and effective:

```js
// each code creates a same tape
tape = new Ciseaux.Sequence().apply("ababab", 1, instruments);
tape = new Ciseaux.Sequence(1, instruments).apply("ababab");
tape = new Ciseaux.Sequence("ababab", instruments).apply(1);

// duplicate arguments are overridden
tape = new Ciseaux.Sequence("ababab", 2, instruments).apply(1);
```

CHANGES:

- `Ciseaux.Sequence#durationPerStep` accepts an array as durationPerStep list.
- `Ciseaux.Sequence#apply()` accepts flexible arguments.

**Comment by s-a** [comment]
I would be glad to try this out. Duration per step are miliseconds?  
Only one thing. Just thinking about if shuffle need negative offset? I think yes. How to declare this case?

**Comment by mohayonao** [comment]
`durationPerStep` is seconds. It is the same as Web Audio API.

I think that supports of negative offset is not need.
Because, `Ciseaux.Sequence` is not for making only shuffle.
If need it, should create an utility function dedicated .

```js
var shuffleSequence = function(pattern, durationPerStep, instruments, offset) {
  var shuffledDurationPerStep = durationPerStep.map(function(durationPerStep, index) {
    return durationPerStep + offset[index % offset.length];
  });
  return new Ciseaux.Sequence(pattern, shuffledDurationPerStep, instruments).apply();
};
```

**Comment by mohayonao** [comment]
implemented: bdf1893382097b37f7a8d8180fb173a415a894bb
